File: kits/comscore-kit/src/main/java/com/mparticle/kits/ComscoreKit.java
package com.mparticle.kits;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;

import com.comscore.analytics.comScore;
import com.mparticle.MPEvent;
import com.mparticle.MParticle;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * <p/>
 * Embedded implementation of the Comscore SDK, tested against Comscore 2.14.0923.
 * <p/>
 */
public class ComscoreKit extends KitIntegration implements KitIntegration.EventListener, KitIntegration.AttributeListener, KitIntegration.ActivityListener {
    //Server config constants defined for this provider
    //keys to provide access to the comscore account.
    private static final String CLIENT_ID = "CustomerC2Value";
    private static final String PUBLISHER_SECRET = "PublisherSecret";
    private static final String USE_HTTPS = "UseHttps";
    private static final String PRODUCT = "product";
    private static final String APPNAME = "appName";
    private static final String AUTOUPDATE_MODE_KEY = "autoUpdateMode";
    private static final String AUTOUPDATE_INTERVAL = "autoUpdateInterval";

    private static final String AUTOUPDATE_MODE_FOREONLY = "foreonly";
    private static final String AUTOUPDATE_MODE_FOREBACK = "foreback";
    private static final String COMSCORE_DEFAULT_LABEL_KEY = "name";
    private String clientId;
    private String publisherSecret;
    private String autoUpdateMode;
    private int autoUpdateInterval = 60;
    private boolean isEnterprise;

    @Override
    public List<ReportingMessage> leaveBreadcrumb(String breadcrumb) {
        return null;
    }

    @Override
    public List<ReportingMessage> logError(String message, Map<String, String> errorAttributes) {
        return null;
    }

    @Override
    public List<ReportingMessage> logException(Exception exception, Map<String, String> exceptionAttributes, String message) {
        return null;
    }

    @Override
    public List<ReportingMessage> logEvent(MPEvent event) {
        if (!isEnterprise) {
            return null;
        }
        List<ReportingMessage> messages = new LinkedList<ReportingMessage>();
        HashMap<String, String> comscoreLabels;
        Map<String, String> attributes = event.getInfo();
        if (attributes == null) {
            comscoreLabels = new HashMap<String, String>();
        }else if (!(attributes instanceof HashMap)){
            comscoreLabels = new HashMap<String, String>();
            for (Map.Entry<String, String> entry : attributes.entrySet())
            {
                comscoreLabels.put(entry.getKey(), entry.getValue());
            }
        }else {
            comscoreLabels = (HashMap<String, String>) attributes;
        }
        comscoreLabels.put(COMSCORE_DEFAULT_LABEL_KEY, event.getEventName());
        if (MParticle.EventType.Navigation.equals(event.getEventType())){
            comScore.view(comscoreLabels);
        }else{
            comScore.hidden(comscoreLabels);
        }
        messages.add(
                ReportingMessage.fromEvent(this,
                        new MPEvent.Builder(event).info(comscoreLabels).build()
                )
        );
        return messages;
    }

    @Override
    public List<ReportingMessage>  logScreen(String screenName, Map<String, String> eventAttributes) {
        return logEvent(
                new MPEvent.Builder(screenName, MParticle.EventType.Navigation)
                        .info(eventAttributes)
                        .build()
        );
    }

    @Override
    public void setUserAttribute(String key, String value) {
        if (isEnterprise){
            comScore.setLabel(KitUtils.sanitizeAttributeKey(key), value);
        }
    }

    @Override
    public void removeUserAttribute(String key) {
        if (isEnterprise){
            comScore.getLabels().remove(KitUtils.sanitizeAttributeKey(key));
        }
    }

    @Override
    public void removeUserIdentity(MParticle.IdentityType identityType) {
        if (isEnterprise){
            comScore.getLabels().remove(identityType.toString());
        }
    }

    @Override
    public List<ReportingMessage> logout() {
        return null;
    }

    @Override
    public void setUserIdentity(MParticle.IdentityType identityType, String id) {
        if (isEnterprise){
            comScore.setLabel(identityType.toString(), id);
        }
    }

    @Override
    public String getName() {
        return "Comscore";
    }

    @Override
    protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
        if (needsRestart()){
            clientId = getSettings().get(CLIENT_ID);
            comScore.setCustomerC2(clientId);
            publisherSecret = getSettings().get(PUBLISHER_SECRET);
            comScore.setPublisherSecret(publisherSecret);
        }

        int tempUpdateInterval = 60;
        try {
            tempUpdateInterval = Integer.parseInt(getSettings().get(AUTOUPDATE_INTERVAL));
        }catch (NumberFormatException nfe){

        }
        if (!getSettings().get(AUTOUPDATE_MODE_KEY).equals(autoUpdateMode) || tempUpdateInterval != autoUpdateInterval){
            autoUpdateInterval = tempUpdateInterval;
            autoUpdateMode = getSettings().get(AUTOUPDATE_MODE_KEY);
            if (AUTOUPDATE_MODE_FOREBACK.equals(autoUpdateMode)){
                comScore.enableAutoUpdate(autoUpdateInterval, false);
            }else if (AUTOUPDATE_MODE_FOREONLY.equals(autoUpdateMode)){
                comScore.enableAutoUpdate(autoUpdateInterval, true);
            }else {
                comScore.disableAutoUpdate();
            }
        }

        boolean useHttps = Boolean.parseBoolean(getSettings().get(USE_HTTPS));
        comScore.setSecure(useHttps);
        comScore.setDebug(getKitManager().getConfigurationManager().getEnvironment() == MParticle.Environment.Development);
        isEnterprise = "enterprise".equals(getSettings().get(PRODUCT));
        String appName = getSettings().get(APPNAME);
        if (appName != null){
            comScore.setAppName(appName);
        }
        return null;
    }

    private boolean needsRestart() {
        return !getSettings().get(CLIENT_ID).equals(clientId) || !getSettings().get(PUBLISHER_SECRET).equals(publisherSecret);
    }

    @Override
    public List<ReportingMessage> onActivityStopped(Activity activity) {
        comScore.onExitForeground();
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(
                new ReportingMessage(this, ReportingMessage.MessageType.APP_STATE_TRANSITION, System.currentTimeMillis(), null)
        );
        return messageList;
    }

    @Override
    public List<ReportingMessage> onActivitySaveInstanceState(Activity activity, Bundle outState) {
        return null;
    }

    @Override
    public List<ReportingMessage> onActivityDestroyed(Activity activity) {
        return null;
    }

    @Override
    public List<ReportingMessage> onActivityCreated(Activity activity, Bundle savedInstanceState) {
        return null;
    }

    @Override
    public List<ReportingMessage> onActivityStarted(Activity activity) {
        comScore.onEnterForeground();
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(
                new ReportingMessage(this, ReportingMessage.MessageType.APP_STATE_TRANSITION, System.currentTimeMillis(), null)
        );
        return messageList;
    }

    @Override
    public List<ReportingMessage> onActivityResumed(Activity activity) {
        return null;
    }

    @Override
    public List<ReportingMessage> onActivityPaused(Activity activity) {
        return null;
    }

    @Override
    public List<ReportingMessage> setOptOut(boolean optOutStatus) {
        comScore.setEnabled(!optOutStatus);
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(
                new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null)
                        .setOptOut(optOutStatus)
        );
        return messageList;
    }

}

File: kits/comscore/comscore-6/src/main/kotlin/com/mparticle/kits/ComscoreKit.kt
package com.mparticle.kits

import android.app.Activity
import android.content.Context
import android.os.Bundle
import com.comscore.Analytics
import com.comscore.PartnerConfiguration
import com.comscore.PublisherConfiguration
import com.comscore.util.log.LogLevel
import com.mparticle.MPEvent
import com.mparticle.MParticle
import com.mparticle.MParticle.IdentityType
import com.mparticle.kits.KitIntegration.ActivityListener
import com.mparticle.kits.KitIntegration.AttributeListener
import com.mparticle.kits.KitIntegration.LogoutListener
import java.util.HashMap
import java.util.LinkedList

class ComscoreKit :
    KitIntegration(),
    KitIntegration.EventListener,
    AttributeListener,
    LogoutListener,
    ActivityListener {
    private var isEnterprise = false

    override fun leaveBreadcrumb(breadcrumb: String): List<ReportingMessage> = emptyList()

    override fun logError(
        message: String,
        errorAttributes: Map<String, String>,
    ): List<ReportingMessage> = emptyList()

    override fun logException(
        exception: Exception,
        exceptionAttributes: Map<String, String>,
        message: String,
    ): List<ReportingMessage> = emptyList()

    override fun logEvent(event: MPEvent): List<ReportingMessage>? {
        if (!isEnterprise) {
            return null
        }
        val messages = LinkedList<ReportingMessage>()
        val comscoreLabels: HashMap<String, String?>
        when (val attributes = event.customAttributeStrings) {
            null -> {
                comscoreLabels = HashMap()
            }
            !is HashMap<String, String?> -> {
                comscoreLabels = HashMap()
                for ((key, value) in attributes) {
                    comscoreLabels[key] = value
                }
            }
            else -> {
                comscoreLabels = attributes
            }
        }
        comscoreLabels[COMSCORE_DEFAULT_LABEL_KEY] = event.eventName
        if (MParticle.EventType.Navigation == event.eventType) {
            Analytics.notifyViewEvent(comscoreLabels)
        } else {
            Analytics.notifyHiddenEvent(comscoreLabels)
        }
        messages.add(
            ReportingMessage.fromEvent(
                this,
                MPEvent.Builder(event).customAttributes(comscoreLabels).build(),
            ),
        )
        return messages
    }

    override fun logScreen(
        screenName: String,
        eventAttributes: Map<String, String>,
    ): List<ReportingMessage> =
        logEvent(
            MPEvent
                .Builder(screenName, MParticle.EventType.Navigation)
                .customAttributes(eventAttributes)
                .build(),
        )!!

    private fun applyEnterpriseScalarUserAttribute(
        key: String,
        value: String,
    ) {
        if (isEnterprise) {
            Analytics
                .getConfiguration()
                .setPersistentLabel(KitUtils.sanitizeAttributeKey(key), value)
        }
    }

    override fun onSetUserAttributeList(
        attributeKey: String?,
        attributeValueList: List<String>?,
        user: FilteredMParticleUser?,
    ) {
        // not supported
    }

    override fun supportsAttributeLists(): Boolean = !isEnterprise

    override fun setAllUserAttributes(
        attributes: Map<String, String>,
        attributeLists: Map<String, List<String>>,
    ) {
        if (isEnterprise) {
            for ((key, value) in attributes) {
                applyEnterpriseScalarUserAttribute(key, value)
            }
        }
    }

    override fun onSetAllUserAttributes(
        userAttributes: Map<String, String>,
        userAttributeLists: Map<String, List<String>>,
        user: FilteredMParticleUser,
    ) {
        setAllUserAttributes(userAttributes, userAttributeLists)
    }

    override fun onRemoveUserAttribute(
        key: String,
        user: FilteredMParticleUser,
    ) {
        if (isEnterprise) {
            Analytics.getConfiguration().removePersistentLabel(KitUtils.sanitizeAttributeKey(key))
        }
    }

    override fun onSetUserAttribute(
        key: String?,
        value: Any?,
        user: FilteredMParticleUser?,
    ) {
        if (key == null || value == null || value !is String) {
            return
        }
        applyEnterpriseScalarUserAttribute(key, value)
    }

    override fun removeUserIdentity(identityType: IdentityType) {
        if (isEnterprise) {
            Analytics.getConfiguration().removePersistentLabel(identityType.toString())
        }
    }

    override fun logout(): List<ReportingMessage> = emptyList()

    override fun setUserIdentity(
        identityType: IdentityType,
        id: String,
    ) {
        if (isEnterprise) {
            Analytics.getConfiguration().setPersistentLabel(identityType.toString(), id)
        }
    }

    override fun getName(): String = NAME

    override fun onKitCreate(
        settings: Map<String, String>,
        context: Context,
    ): List<ReportingMessage> {
        val partnerConfiguration =
            PartnerConfiguration
                .Builder()
                .partnerId(settings[PARTNER_ID])
                .build()
        Analytics.getConfiguration().addClient(partnerConfiguration)
        val builder = PublisherConfiguration.Builder()
        builder.publisherId(getSettings()[CLIENT_ID])
        builder.secureTransmission(true)
        if (MParticle.getInstance()?.environment == MParticle.Environment.Development) {
            Analytics.setLogLevel(LogLevel.VERBOSE)
        }
        isEnterprise = ENTERPRISE_STRING == getSettings()[PRODUCT]
        val publisherConfiguration = builder.build()
        Analytics.getConfiguration().addClient(publisherConfiguration)
        Analytics.start(context)
        return emptyList()
    }

    override fun onActivityPaused(activity: Activity): List<ReportingMessage> {
        Analytics.notifyExitForeground()
        val messageList = LinkedList<ReportingMessage>()
        messageList.add(
            ReportingMessage(
                this,
                ReportingMessage.MessageType.APP_STATE_TRANSITION,
                System.currentTimeMillis(),
                null,
            ),
        )
        return messageList
    }

    override fun onActivitySaveInstanceState(
        activity: Activity,
        outState: Bundle?,
    ): List<ReportingMessage> = emptyList()

    override fun onActivityDestroyed(activity: Activity): List<ReportingMessage> = emptyList()

    override fun onActivityCreated(
        activity: Activity,
        savedInstanceState: Bundle?,
    ): List<ReportingMessage> = emptyList()

    override fun onActivityResumed(activity: Activity): List<ReportingMessage> {
        Analytics.notifyEnterForeground()
        val messageList = LinkedList<ReportingMessage>()
        messageList.add(
            ReportingMessage(
                this,
                ReportingMessage.MessageType.APP_STATE_TRANSITION,
                System.currentTimeMillis(),
                null,
            ),
        )
        return messageList
    }

    override fun onActivityStarted(activity: Activity): List<ReportingMessage> = emptyList()

    override fun onActivityStopped(activity: Activity): List<ReportingMessage> = emptyList()

    override fun setOptOut(optOutStatus: Boolean): List<ReportingMessage> {
        if (!optOutStatus) {
            Analytics.getConfiguration().disable()
        }
        val messageList = LinkedList<ReportingMessage>()
        messageList.add(
            ReportingMessage(
                this,
                ReportingMessage.MessageType.OPT_OUT,
                System.currentTimeMillis(),
                null,
            ).setOptOut(optOutStatus),
        )
        return messageList
    }

    companion object {
        private const val CLIENT_ID = "CustomerC2Value"
        private const val PRODUCT = "product"
        private const val PARTNER_ID = "partnerId"
        private const val COMSCORE_DEFAULT_LABEL_KEY = "name"
        private const val ENTERPRISE_STRING = "enterprise"
        const val NAME = "Comscore"
    }
}
